### ISSUE TITLE
Very confused about the tables 

### ISSUE THREAD

**Issue by julie1013** [ga-wdi-boston/full-stack-project#408]
Again, relationship is user has many books and books has many users. Through the join table qualified_books as per Chris's suggestion. This is essentially a to-read list of books that haven't been deemed in need of removal yet. But is books a master list? Is it an individual list? What about the join table? It's essentially a to-read list and Everyone should have their own. But can you do that with a join table? I don't know. I'm extraordinary confused about this and need direction in sorting it out. 

Thanks

**Comment by berziiii** [comment]
Join tables allows a user to choose the books they want to read. In a sense, the books table is the "master list", as you put it, that holds all books that any user can see. 

Where a user will own their data is that they can only remove/update books they've added to the book collection and manage the books in their list of `qualified_books` list.

**Comment by julie1013** [comment]
Thanks. At the gym now but I'm getting to GA no later than 4 to work.

**Comment by payne-chris-r** [comment]
Think about what data you're storing in a join table. What are your columns in the qualified_books table?

**Comment by julie1013** [comment]
Chris: They're `user_id` and `book_id`

Brian: What do you mean "their list of `qualified_books` list? Does this mean that each user has their own separate `qualified_books` to link the user to the books master list?

**Comment by julie1013** [comment]
So if I wanted a user to see their PERSONAL to-read list upon login, I would essentially have to display a result of grabbing the user's `user_id`  and the associated `books` from `qualified_books`?

**Comment by berziiii** [comment]
Have you tried seeding data? Set up your relationships and add data to users, books and qualified books and see what you see. this will help you see how the join table works and how data will be presented

**Comment by julie1013** [comment]
Yeah, I have all that stuff set up. I guess at this point I just don't "feel" it the way I do with JavaScript and other programming I'm familiar with. I'm not sure how to HANDLE it at this point, if that makes sense.

**Comment by julie1013** [comment]
I am still very confused and need at least a hint where to go from here. I haven't written one line of code today. Not one.

**Comment by julie1013** [comment]
Can I request a one-on-one for Monday?

**Comment by berziiii** [comment]
Oh just said that you have data seeded. If this is the case, you should be able to execute a curl request to add a book to a users list since you have be id of the user and book.

**Comment by berziiii** [comment]
From this, you should be able to create a form in the client to fire the same request. Which adds a book to a users book list.

**Comment by julie1013** [comment]
So would a user's list be a new table?

**Comment by berziiii** [comment]
No. The users list of books is the result of the join table called `qualified books`. If you have data seeded and fire a GET request to /books, you'll see which users a book has a relationship with. You can also fire a GET request to a user, you'll need a valid token, and if your serializer for a user has the attribute `:books` on it then you'll see the books that the user has, which is the result of the join table. 

The join is acting as the 'connecter' of a user and books. When the relationships are set up correctly, you don't need to search, or even touch the join table, to see the relationship that the join is responsible for. In your case, it's a user has many books. And if you look at a specific book, a book may have many users.

**Comment by julie1013** [comment]
I can only see which users are associated with which books on the join table... should I be able to see users from the books table?

**Comment by payne-chris-r** [comment]
That's the point of a join table. You may want to go back and look at the sql lessons on join tables or look at some of the examples we did with many-to-many. Think about the relationships we used in those lessons and how they might apply to your problem. I also agree with Brian that writing curl requests that POST to your join and GET things out of your join table, you may learn a thing or two about what purpose your join table serves.

**Comment by julie1013** [comment]
I'm sorry. I don't understand. At all. I need a one-on-one tomorrow. I haven't gotten anything accomplished this weekend.

**Comment by payne-chris-r** [comment]
Did you go back and look at the rails lessons for many-to-many? Doctors have many Patients through appointments and Patients have many Doctors through appointments. Can you think of what you might say about your relationship??

**Comment by julie1013** [comment]
Yes, several times. It's just that I don't know what to do once the tables are there. Something just isn't clicking.

**Comment by payne-chris-r** [comment]
Ok. Can you write it here?

**Comment by payne-chris-r** [comment]
Once you've done that, in your own words, what is the reason for a join table?

**Comment by julie1013** [comment]
Write what?

**Comment by payne-chris-r** [comment]
What your relationships are

**Comment by julie1013** [comment]
User has many books, books has many users. qualified_books references both entities with user_id and book_id.

**Comment by julie1013** [comment]
Join table purpose is to reference both entities as it is a many-to-many relationship. It is necessary as the relationship is reciprocal, for lack of a better term, unlike a one to many relationship

**Comment by payne-chris-r** [comment]
That's exactly right. So what does a single record in your join table look like in SQL? What does it represent?

**Comment by julie1013** [comment]
It is pointing to specific entries in the other tables.

**Comment by payne-chris-r** [comment]
So what does a single record in your join table look like in SQL? What does it represent?

**Comment by julie1013** [comment]
```
{
"id": 1,
"user": {
"id": 1,
"email": "dog@dog.com",
"books": [
{
"id": 2,
"title": "Mercy",
"author": "Jodi Picoult",
"created_at": "2016-09-30T19:52:09.888Z",
"updated_at": "2016-09-30T19:52:09.888Z"
},
{
"id": 3,
"title": "The Da Vinci Code",
"author": "Dan Brown",
"created_at": "2016-09-30T19:52:42.701Z",
"updated_at": "2016-09-30T19:53:11.515Z"
}
]
},
"book": {
"id": 2,
"title": "Mercy",
"author": "Jodi Picoult",
"created_at": "2016-09-30T19:52:09.888Z",
"updated_at": "2016-09-30T19:52:09.888Z"
}
},
```

This represents the user and their books.

**Comment by payne-chris-r** [comment]
Ok. You're absolutely right. What are "their books" in relation to your app?

**Comment by julie1013** [comment]
What do you mean?

**Comment by payne-chris-r** [comment]
How would you explain to someone in laymans terms (pretend you're talking to your mom/grandma) what single records in that join table represent?

**Comment by payne-chris-r** [comment]
How many users do you currently have in your db?

**Comment by julie1013** [comment]
Each record represents a user and a book associated with that user.
I think I have two or three users.

**Comment by payne-chris-r** [comment]
Not exactly. What did a record in the "appointments" table represent for doctors and patients?

**Comment by julie1013** [comment]
What doctor was going to see what patient.

**Comment by julie1013** [comment]
Or what doctor was associated with which patient.

**Comment by payne-chris-r** [comment]
Yep. Exactly. So each record in your join represents what?

**Comment by julie1013** [comment]
What book is associated with which user.

**Comment by payne-chris-r** [comment]
Exactly!!

**Comment by julie1013** [comment]
Yeah... and I'm not sure where to go from there.

**Comment by payne-chris-r** [comment]
You have a table full of users, a table full of books, and a table full of different "books that belong to a specific user." How would I see that entire table in SQL? How would I see it in the browser?

**Comment by julie1013** [comment]
`http://localhost:3000/users/1/books` ? Except it shows all the books, not just the ones I gave the user.

**Comment by payne-chris-r** [comment]
Think about the curl request you might make. Were we able to make a request to get all appointments for the clinic?

**Comment by julie1013** [comment]
You mean this? `curl --include --request GET http://localhost:3000/qualified_books`

```
{"qualified_books":[{"id":1,"user":{"id":1,"email":"dog@dog.com","books":[{"id":2,"title":"Mercy","author":"Jodi Picoult","created_at":"2016-09-30T19:52:09.888Z","updated_at":"2016-09-30T19:52:09.888Z"},{"id":3,"title":"The Da Vinci Code","author":"Dan Brown","created_at":"2016-09-30T19:52:42.701Z","updated_at":"2016-09-30T19:53:11.515Z"}]},"book":{"id":2,"title":"Mercy","author":"Jodi Picoult","created_at":"2016-09-30T19:52:09.888Z","updated_at":"2016-09-30T19:52:09.888Z"}},{"id":2,"user":{"id":1,"email":"dog@dog.com","books":[{"id":2,"title":"Mercy","author":"Jodi Picoult","created_at":"2016-09-30T19:52:09.888Z","updated_at":"2016-09-30T19:52:09.888Z"},{"id":3,"title":"The Da Vinci Code","author":"Dan Brown","created_at":"2016-09-30T19:52:42.701Z","updated_at":"2016-09-30T19:53:11.515Z"}]},"book":{"id":3,"title":"The Da Vinci Code","author":"Dan Brown","created_at":"2016-09-30T19:52:42.701Z","updated_at":"2016-09-30T19:53:11.515Z"}}]}
```

**Comment by payne-chris-r** [comment]
Exactly! Can you post your serializer for `qualified_books`?

**Comment by julie1013** [comment]
``` ruby
class QualifiedBookSerializer < ActiveModel::Serializer
  attributes :id
  has_one :user
  has_one :book
end
```

**Comment by julie1013** [comment]
Ok, so I have SOMETHING going, using what we did with tic-tac-toe as a reference. I have this piece of code:

``` js
const showMyToReadList = function () {
  let userID = app.user.id;
  return $.ajax({
    url: app.host + '/users/' + userID + '/books',
    method: 'GET',
    headers: {
      Authorization: 'Token token=' + app.user.token,
    },
  });
};
```

And that returns ALL the books, not the books belonging to one user. I'm guessing it might be something in the controller? But I'm not sure about that.

**Comment by payne-chris-r** [comment]
Ok. I don't think you necessarily want to serialize out "has_many :user" and "has_many :book". You might, but I typically just serialize out the user_id and book_id until I need the whole thing.

**Comment by payne-chris-r** [comment]
As far as your Ajax function is concerned, that looks great. You need to walk the request all the way through your backend and figure out what it's actually returning. Is returning ALL the books what you want a GET to users/:id/books to return? Where do we start looking on the backend?

**Comment by payne-chris-r** [comment]
Also, I recommend trying it with curl before you figure it out in js/ajax!

**Comment by julie1013** [comment]
...Isn't it serializing out has_one :user and has_one :book? Where is it doing has_many?

**Comment by julie1013** [comment]
I tried this curl request:

```
curl --include --request GET http://localhost:3000/qualified_books \
    --header "Content-Type: application/json" \
    --data '{
      "qualified_book": {
        "user_id": "1"
      }
    }'
```

I get this:

```
// edited by @jrhorn424 for readability using the `jq` command line app
{{
  "qualified_books": [
    {
      "id": 1,
      "user": {
        "id": 1,
        "email": "dog@dog.com",
        "books": [
          {
            "id": 2,
            "title": "Mercy",
            "author": "Jodi Picoult",
            "created_at": "2016-09-30T19:52:09.888Z",
            "updated_at": "2016-09-30T19:52:09.888Z"
          },
          {
            "id": 3,
            "title": "The Da Vinci Code",
            "author": "Dan Brown",
            "created_at": "2016-09-30T19:52:42.701Z",
            "updated_at": "2016-09-30T19:53:11.515Z"
          }
        ]
      },
      "book": {
        "id": 2,
        "title": "Mercy",
        "author": "Jodi Picoult",
        "created_at": "2016-09-30T19:52:09.888Z",
        "updated_at": "2016-09-30T19:52:09.888Z"
      }
    },
    {
      "id": 2,
      "user": {
        "id": 1,
        "email": "dog@dog.com",
        "books": [
          {
            "id": 2,
            "title": "Mercy",
            "author": "Jodi Picoult",
            "created_at": "2016-09-30T19:52:09.888Z",
            "updated_at": "2016-09-30T19:52:09.888Z"
          },
          {
            "id": 3,
            "title": "The Da Vinci Code",
            "author": "Dan Brown",
            "created_at": "2016-09-30T19:52:42.701Z",
            "updated_at": "2016-09-30T19:53:11.515Z"
          }
        ]
      },
      "book": {
        "id": 3,
        "title": "The Da Vinci Code",
        "author": "Dan Brown",
        "created_at": "2016-09-30T19:52:42.701Z",
        "updated_at": "2016-09-30T19:53:11.515Z"
      }
    },
    {
      "id": 3,
      "user": {
        "id": 1,
        "email": "dog@dog.com",
        "books": [
          {
            "id": 2,
            "title": "Mercy",
            "author": "Jodi Picoult",
            "created_at": "2016-09-30T19:52:09.888Z",
            "updated_at": "2016-09-30T19:52:09.888Z"
          },
          {
            "id": 3,
            "title": "The Da Vinci Code",
            "author": "Dan Brown",
            "created_at": "2016-09-30T19:52:42.701Z",
            "updated_at": "2016-09-30T19:53:11.515Z"
          }
        ]
      },
      "book": null
    }
  ]
}
```

I'm not sure why it's returning the same user more than once.

**Comment by jrhorn424** [comment]
> I'm not sure why it's returning the same user more than once.

Because one user has more than one book they want to read. Your serializer is confusing. Try making the change @payne-chris-r requested:

> ... but I typically just serialize out the user_id and book_id until I need the whole thing.

Your curl request looks like:

```
curl --include --request GET http://localhost:3000/qualified_books
```

But your pasted AJAX as this URL:

```
    url: app.host + '/users/' + userID + '/books',
```

So I'm not sure your AJAX is doing what you think it is.

**Comment by jrhorn424** [comment]
What view are you trying to render? The "To Read" list? If so, you should be able to make a request to `/users/:id/books` and get what you want. I think some of the confusion here is that you are making GET requests to the JOIN resource, which typically isn't done. You'll make POST requests there, for sure, but not GET requests. GET requests should be done through a primary resource instead of the JOIN resource.

**Comment by jrhorn424** [comment]
Try making the output for your curl request to `qualified_books` look like this by changing the serializer:

```
{
  "qualified_books": [
    {
      "id": 1,
      "user_id": 1,
      "book_id": 2
    },
    {
      "id": 2,
      "user_id": 1,
      "book_id": 3
    },
    {
      "id": 3,
      "user_id": 1,
      "book_id": null
    }
  ]
}
```

**Comment by julie1013** [comment]
You mean take away this?:

``` ruby
  has_one :user
  has_one :book
```

**Comment by julie1013** [comment]
Also, problem is that this, for example, returns all of the books instead of just the books that the user is associated with: `curl --include --request GET http://localhost:3000/users/1/books`

**Comment by jrhorn424** [comment]
You're working on two things at once. Just get your request to qualified books to work as I suggested, then I'll answer your concern about the nested resource.

**Comment by julie1013** [comment]
Already have. Changing the serializer gets this: 

```
{
"qualified_books": [
{
"id": 1,
"user_id": 1,
"book_id": 2
},
{
"id": 2,
"user_id": 1,
"book_id": 3
},
{
"id": 3,
"user_id": 1,
"book_id": null
}
]
}
```

And I made a loop to bring the to-read list (in a very primitive form, obviously, but that can easily be fixed... and of course this still has the issue of it returning all the books instead of just the ones associated with the user). This just from books and user:

``` js
const onShowMyToReadListSuccess = function (data){
  for(let i = 0; i < data.length; i++){
    $('.to-read').append(data[i].title);
  }
};
```

**Comment by julie1013** [comment]
Anyway, I'm going to bed now. Thanks for all your help-- you guys are awesome!

I'd still like a one-on-one tomorrow if possible.

Thanks again,

J

**Comment by julie1013** [comment]
So how's about the nested resource?

**Comment by jrhorn424** [comment]
> So how's about the nested resource?

Now it's a controller concern, as you guessed earlier. :smile:

In the appropriate action in your books controller you'll need to use an if statement to check whether `params[:user_id]` is present, and if it is, filter by user, else return all books. If you need to double check where to put this logic, ensure you have the nested route defined and look for the output in `rake routes`.

**Comment by julie1013** [comment]
``` ruby
 def show
    if params[:user_id].present?
      render json: Book.find(params[:user_id])
    else render json: Book.find(params[:id])
    end
  end
```

Is this even remotely on the right track? ;)

**Comment by jrhorn424** [comment]
Yeah, I don't think `show` is the correct action. Check `rake routes` and paste the output here.

Also, this is a bit more idiomatic for Rails:

``` rb
 def show # probably not what you want
  if params[:user_id].present?
    user = params[:user_id]
    render json: Book.find_by(user: user)
  else 
    render json: Book.find(params[:id]) # probably not what you want, related to being in the wrong action
  end
end
```

**Comment by julie1013** [comment]
Not show... should it be a new user controller action then?

**Comment by payne-chris-r** [comment]
Did you figure this out @julie1013 ? If so, can you post a solution and close? Thanks!

**Comment by julie1013** [comment]
Not yet... I was cheering because I figured out something else!

**Comment by jrhorn424** [comment]
No, it isn't a concern of the user controller. Post your `rake routes` output, as I've asked a few times.

**Comment by julie1013** [comment]
```
  Prefix Verb   URI Pattern                              Controller#Action
qualified_books GET    /qualified_books(.:format)               qualified_books#index
                POST   /qualified_books(.:format)               qualified_books#create
 qualified_book GET    /qualified_books/:id(.:format)           qualified_books#show
                PATCH  /qualified_books/:id(.:format)           qualified_books#update
                PUT    /qualified_books/:id(.:format)           qualified_books#update
                DELETE /qualified_books/:id(.:format)           qualified_books#destroy
       examples GET    /examples(.:format)                      examples#index
                POST   /examples(.:format)                      examples#create
        example GET    /examples/:id(.:format)                  examples#show
                PATCH  /examples/:id(.:format)                  examples#update
                PUT    /examples/:id(.:format)                  examples#update
                DELETE /examples/:id(.:format)                  examples#destroy
        sign_up POST   /sign-up(.:format)                       users#signup
        sign_in POST   /sign-in(.:format)                       users#signin
                DELETE /sign-out/:id(.:format)                  users#signout
                PATCH  /change-password/:id(.:format)           users#changepw
     user_books GET    /users/:user_id/books(.:format)          books#index
                POST   /users/:user_id/books(.:format)          books#create
  new_user_book GET    /users/:user_id/books/new(.:format)      books#new
 edit_user_book GET    /users/:user_id/books/:id/edit(.:format) books#edit
      user_book GET    /users/:user_id/books/:id(.:format)      books#show
                PATCH  /users/:user_id/books/:id(.:format)      books#update
                PUT    /users/:user_id/books/:id(.:format)      books#update
                DELETE /users/:user_id/books/:id(.:format)      books#destroy
          users GET    /users(.:format)                         users#index
           user GET    /users/:id(.:format)                     users#show
          books GET    /books(.:format)                         books#index
                POST   /books(.:format)                         books#create
           book GET    /books/:id(.:format)                     books#show
                PATCH  /books/:id(.:format)                     books#update
                PUT    /books/:id(.:format)                     books#update
                DELETE /books/:id(.:format)                     books#destroy
        reviews GET    /reviews(.:format)                       reviews#index
                POST   /reviews(.:format)                       reviews#create
         review GET    /reviews/:id(.:format)                   reviews#show
                PATCH  /reviews/:id(.:format)                   reviews#update
                PUT    /reviews/:id(.:format)                   reviews#update
                DELETE /reviews/:id(.:format)                   reviews#destroy
```

**Comment by jrhorn424** [comment]
> `user_books GET    /users/:user_id/books(.:format)          books#index`

**Comment by jrhorn424** [comment]
What controller action should it be?

**Comment by julie1013** [comment]
books.

``` ruby
  def index
    @books = Book.all
    if params[:user_id].present?
      user = params[:user_id]
      render json: Book.find_by(user: user)
    else
    render json: @books
    end
  end
```

**Comment by payne-chris-r** [comment]
That's correct. So what are the current symptoms? What's the problem you're trying to solve?

**Comment by julie1013** [comment]
OK, so I should put this in the books controller, correct?

Trying to get it so that when I do this URL: `http://localhost:3000/users/1/books`
(or users/2/ or whatever) that it returns ONLY the books that the person has on their list. It's showing all the books that are in the database. After I change the index method, I get this:

`ActiveRecord::StatementInvalid in BooksController#index
PG::UndefinedColumn: ERROR: column books.user does not exist LINE 1: SELECT "books".* FROM "books" WHERE "books"."user" = '1' LI... ^ : SELECT "books".* FROM "books" WHERE "books"."user" = '1' LIMIT 1`

```Extracted source (around line #7):
5
6
7
8
9
10

``` ruby
    if params[:user_id].present?
      user = params[:user_id]
      render json: Book.find_by(user: user)
    else
      render json: @books
    end
```

line 7 is `render json: Book.find_by(user: user)`

**Comment by payne-chris-r** [comment]
The pertinent line from that error message is this:

```
ERROR: column books.user does not exist
```

**Comment by julie1013** [comment]
Yess.... But why does it need to? As I keep saying, I'm not super clear on the tables. :(

**Comment by payne-chris-r** [comment]
Why does it need to, what?

**Comment by julie1013** [comment]
eeegggh I mean have that column? What's it for? Or is it referring to my join table which does not have that name?

**Comment by jrhorn424** [comment]
What is your model code for `Book`, `QualifiedBook`, and `User`?

**Comment by julie1013** [comment]
Book:

``` ruby
class Book < ActiveRecord::Base
  has_many :reviews
  has_many :users, through: :qualified_books
  has_many :qualified_books, dependent: :destroy
end
```

QualifiedBook

``` ruby
class QualifiedBook < ActiveRecord::Base
  belongs_to :user, inverse_of: :qualified_books
  belongs_to :book, inverse_of: :qualified_books
end
```

User

``` ruby
class User < ActiveRecord::Base
  include Authentication
  has_many :reviews
  has_many :books, through: :qualified_books
  has_many :qualified_books, dependent: :destroy
end
```

**Comment by payne-chris-r** [comment]
Do you know what this line is doing?

``` ruby
Book.find_by(user: user)
```

**Comment by julie1013** [comment]
Find book by user that the book is associated with.

**Comment by jrhorn424** [comment]
What does the output of `find_by` give us? I think I may have lead you astray. Can you find a better method than `find_by` that does what we want? I think if you find the return value of `find_by` you'll notice my mistake.

**Comment by julie1013** [comment]
Judging from the error message, it gives `undefined`. I think. Because it's targeting "books"."user" which does not exist. 

Would it be `User.find(user_id)`?

**Comment by jrhorn424** [comment]
> Would it be User.find(user_id)?

Does that return users or books? Don't guess. `Book` isn't wrong. In my last message, I asked you to find something other than `find_by`. `find_by` is the problem.

**Comment by julie1013** [comment]
Yeah, I know, I looked stuff up and it looked like it should perhaps be `find`.
I tried this:

``` ruby
  def index
    @books = Book.all
    if params[:user_id].present?
      user = params[:user_id]
      render json: Book.find(user)
    else
      render json: @books
    end
  end
```

And it finds the book, using the user's ID as the book ID. Welll, that makes sense. It's only passing in a number. Nothing more specific than that. Meantime I'm trying to look up other rails methods...

**Comment by jrhorn424** [comment]
How many books should we get back?

**Comment by julie1013** [comment]
Well, I want it to get ALL of the user's books but it looks like it's taking the user ID and interpreting it as a book ID. So if I pass 1, I get back the book with the ID of 1.

**Comment by jrhorn424** [comment]
That's right. `Book.find` gets a single book. You want to get more than one book. What ActiveRecord method can we call on `Book` to get books that match a set criteria?

I believe you want `Book.where(user: user)`.

**Comment by julie1013** [comment]
That was actually the last thing I tried, but it still didn't work.

Code:

``` ruby
  def index
    @books = Book.all
    if params[:user_id]
      user = params[:user_id]
      render json: Book.where(user: user)
    else
      render json: @books
    end
  end
```

This error:

```
PG::UndefinedColumn: ERROR: column books.user does not exist LINE 1: SELECT "books".* FROM "books" WHERE "books"."user" = '1' ^ : SELECT "books".* FROM "books" WHERE "books"."user" = '1'
```

**Comment by jrhorn424** [comment]
My mistake.

I'd try the following next: `Book.where(users: [user])`.

If that doesn't work, try: `Books.where(id: QualifiedBook.where(user: user))`

Finally, if that doesn't work, try: `Books.where(id: QualifiedBook.where(user: user).pluck(:id))`

@gaand Can you comment on what is conventional for loading a subset of a join? I think I prefer the second or third.

Also, please restructure your code (this is a little more rails-y):

``` rb
  def index
    user = params[:user_id]
    @books = if user
               # whatever works from the above suggestions
             else
               Book.all
             end
    render json: @books
  end
```

**Comment by payne-chris-r** [comment]
def index
    user = params[:user_id]
    @books = if user
               # whatever works from the above suggestions
             else
               Book.all
             end
    render json: @books
  end

**Comment by payne-chris-r** [comment]
Woops. Clicked "close and comment" thinking it was `Cancel` again... classic!

**Comment by julie1013** [comment]
Unfortunately, none of these work. I got this error for all of them:

`No route matches [GET] "/users/1/books"`

**Comment by jrhorn424** [comment]
I clearly see this in your previous rake routes output:

> `user_books GET    /users/:user_id/books(.:format)          books#index`

That looks like a `[GET] "/users/1/books"` to me. Did you change your routes?

Please paste in this issue the output of `bundle exec rake routes | grep books`.

**Comment by julie1013** [comment]
OK, I had this (per my brother's suggestion) from a couple days ago:

``` ruby
resources :users, only: [:index, :show, :signup, :signout, :changepw] do
    resources :books
  end
```

Yes, I did remove it. But I also put it back, just now. I still get errors, albeit different ones (and I had already tried this with that bit of code in anyway):

For `Book.where(users: [user])`
`ActiveRecord::StatementInvalid in BooksController#index`
`PG::UndefinedColumn: ERROR: column books.user_id does not exist LINE 1: SELECT "books".* FROM "books" WHERE "books"."user_id" = '1' ^ : SELECT "books".* FROM "books" WHERE "books"."user_id" = '1'`

For `Books.where(id: QualifiedBook.where(user: user))`
`NameError in BooksController#index`
`uninitialized constant BooksController::Books`

For `Books.where(id: QualifiedBook.where(user: user).pluck(:id))`
`NameError in BooksController#index`
`uninitialized constant BooksController::Books`

As of now, here's `bundle exec rake routes | grep books`:

`qualified_books GET    /qualified_books(.:format)     qualified_books#index
                POST   /qualified_books(.:format)     qualified_books#create
 qualified_book GET    /qualified_books/:id(.:format) qualified_books#show
                PATCH  /qualified_books/:id(.:format) qualified_books#update
                PUT    /qualified_books/:id(.:format) qualified_books#update
                DELETE /qualified_books/:id(.:format) qualified_books#destroy
          books GET    /books(.:format)               books#index
                POST   /books(.:format)               books#create
           book GET    /books/:id(.:format)           books#show
                PATCH  /books/:id(.:format)           books#update
                PUT    /books/:id(.:format)           books#update
                DELETE /books/:id(.:format)           books#destroy
~/WDI_Fall_2016/projects/book-sucks/book-sucks-backend (booksucks)$ bundle exec rake routes | grep books
qualified_books GET    /qualified_books(.:format)               qualified_books#index
                POST   /qualified_books(.:format)               qualified_books#create
 qualified_book GET    /qualified_books/:id(.:format)           qualified_books#show
                PATCH  /qualified_books/:id(.:format)           qualified_books#update
                PUT    /qualified_books/:id(.:format)           qualified_books#update
                DELETE /qualified_books/:id(.:format)           qualified_books#destroy
     user_books GET    /users/:user_id/books(.:format)          books#index
                POST   /users/:user_id/books(.:format)          books#create
  new_user_book GET    /users/:user_id/books/new(.:format)      books#new
 edit_user_book GET    /users/:user_id/books/:id/edit(.:format) books#edit
      user_book GET    /users/:user_id/books/:id(.:format)      books#show
                PATCH  /users/:user_id/books/:id(.:format)      books#update
                PUT    /users/:user_id/books/:id(.:format)      books#update
                DELETE /users/:user_id/books/:id(.:format)      books#destroy
          books GET    /books(.:format)                         books#index
                POST   /books(.:format)                         books#create
           book GET    /books/:id(.:format)                     books#show
                PATCH  /books/:id(.:format)                     books#update
                PUT    /books/:id(.:format)                     books#update
                DELETE /books/:id(.:format)                     books#destroy
~/WDI_Fall_2016/projects/book-sucks/book-sucks-backend (booksucks)$ bundle exec rake routes | grep books
qualified_books GET    /qualified_books(.:format)               qualified_books#index
                POST   /qualified_books(.:format)               qualified_books#create
 qualified_book GET    /qualified_books/:id(.:format)           qualified_books#show
                PATCH  /qualified_books/:id(.:format)           qualified_books#update
                PUT    /qualified_books/:id(.:format)           qualified_books#update
                DELETE /qualified_books/:id(.:format)           qualified_books#destroy
     user_books GET    /users/:user_id/books(.:format)          books#index
                POST   /users/:user_id/books(.:format)          books#create
  new_user_book GET    /users/:user_id/books/new(.:format)      books#new
 edit_user_book GET    /users/:user_id/books/:id/edit(.:format) books#edit
      user_book GET    /users/:user_id/books/:id(.:format)      books#show
                PATCH  /users/:user_id/books/:id(.:format)      books#update
                PUT    /users/:user_id/books/:id(.:format)      books#update
                DELETE /users/:user_id/books/:id(.:format)      books#destroy
          books GET    /books(.:format)                         books#index
                POST   /books(.:format)                         books#create
           book GET    /books/:id(.:format)                     books#show
                PATCH  /books/:id(.:format)                     books#update
                PUT    /books/:id(.:format)                     books#update
                DELETE /books/:id(.:format)                     books#destroy`

**Comment by gaand** [comment]
Is this a place to use `current_user.books`?  At least for a user getting a list of their books?

**Comment by jrhorn424** [comment]
@julie1013 Great. Thanks for the output.

I made (yet another) mistake. I'm a bit surprised you didn't call me on it! There's a spelling error in two of my above suggestions. It's a pluralization mistake. Can you find it?

Also, I do recommend following @gaand's suggestion of trying `current_user.books`.

I'm closing this issue since we've already solved several issues. The length and breadth of this issue makes it difficult as a resource to learn from.

Please open a new issue and format your post correctly.

If you're confused about how to format your post, please review other posts where I've asked you or others to fix your formatting. (You can search for "format" in just about any tracker.)

**Comment by julie1013** [comment]
This works!!!

`Book.where(id: QualifiedBook.where(user: user).pluck(:id))`

**Comment by jrhorn424** [comment]
@julie1013 That works for apps where you need to see other user's lists. Is that the use case for this app?

**Comment by julie1013** [comment]
I don't think it's very important for these purposes whether or not others can see the user's lists. The idea is, that since this is a simple app, the user picks bad books from a master list, adds them to to-read list if they want to, and has the option of writing a review.

**Comment by jrhorn424** [comment]
Cool. So my suggestion of `Book.where(id: QualifiedBook.where(user: user).pluck(:id))` is what I've used in the past when other people need to see my lists.

If I'm the only one that needs to see my list, its probably better to make `BooksController < ProtectedController` and then use `current_user.books`.

**Comment by julie1013** [comment]
Sounds good! At long last, issue solved!
